### PR TITLE
Add kinodynamical models for robots

### PR DESCRIPTION
**Pull request by edantec**
This PR aims at adding a kinodynamics forward scheme to Aligator. This kinodynamics scheme takes into account the full kinematics of the robot and its centroidal dynamics expressed in base frame. The derivatives are easier to compute than the full dynamics ones, at the price of having to rely on an inverse dynamics QP to output the commanded torques. 

A Python script of an optimal control problem involving Solo walking is provided as an example.

### CODE REVIEW

File: include/aligator/modelling/dynamics/kinodynamics-fwd.hxx [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+          (pdata.Jcom - d.fJf_.template topRows<3>());
+    }
+  }
+  // Compute d(Ag * q_ddot)/ dq
+  d.a0_.template head<6>().setZero();
+  d.a0_.tail(pin_model_.nv - 6) = a;
+  pinocchio::computeCentroidalDynamicsDerivatives(pin_model_, pdata, q, d.v0_,
+                                                  d.a0_, d.dh_dq_, d.dhdot_dq_,
+                                                  d.dhdot_dv_, d.dhdot_da_);
+  d.Jx_.block(pin_model_.nv, 0, 6, pin_model_.nv) -= d.Agu_inv_ * d.dhdot_dq_;
```
Comment (jcarpent) [inline include/aligator/modelling/dynamics/kinodynamics-fwd.hxx:126]: Use noalias when doing such computations:
```suggestion
  d.Jx_.block(pin_model_.nv, 0, 6, pin_model_.nv).noalias() -= d.Agu_inv_ * d.dhdot_dq_;
```

File: include/aligator/modelling/dynamics/kinodynamics-fwd.hxx [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+
+  // Compute d(Ag_dot * q_dot)/ dq
+  d.a0_.setZero();
+  pinocchio::computeCentroidalDynamicsDerivatives(pin_model_, pdata, q, v,
+                                                  d.a0_, d.dh_dq_, d.dhdot_dq_,
+                                                  d.dhdot_dv_, d.dhdot_da_);
+  d.Jx_.block(pin_model_.nv, 0, 6, pin_model_.nv) -= d.Agu_inv_ * d.dhdot_dq_;
```
Comment (jcarpent) [inline include/aligator/modelling/dynamics/kinodynamics-fwd.hxx:133]: same here

File: include/aligator/modelling/dynamics/kinodynamics-fwd.hxx [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+
+  // Compute d(Ag_dot * q_dot)/ dq_dot
+  d.Jx_.block(pin_model_.nv, pin_model_.nv, 6, pin_model_.nv) =
```
Comment (jcarpent) [inline include/aligator/modelling/dynamics/kinodynamics-fwd.hxx:136]: same here

**Review by jcarpent — CHANGES_REQUESTED** [review]
Thanks @edantec for these nice contributions.
I provide minor comments.

Overall, the computational efficiency can be improved but it can wait.

**Comment by ManifoldFR** [comment]
@jcarpent do you still need to reviesw this PR?

**Pull request merged**